cpp/include/cudf/join_new.hpp

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+namespace join {
+
+struct IdentityFilter {
```

Comment:
So one of the drawbacks to the filter approach is that we need to put more of the logic into header files in order to pass this filter as a template parameter.  Also, when we get to implementation, it would be nice to SKIP the Identify Filter.  That is, at some point we're going to iterate over an intermediate result and apply this filter.  If this filter is not going to modify the intermediate result in any way (as the IdentifyFilter would) it would be more efficient not even to launch the kernel.